Adds support for GitLabAPI V4 while keeping support for V3

Argelbargel opened the pull request:
This is an enhancement of PR #596 which adds support for gitlab-api v4 while keeping support for v3
(as i forked jenkins/gitlab-plugin i could not find a simple way to add to @ZJvandeWeg 's PR). 
GitLabConnectionConfig has a new configuration option "API-Level" in the advanced configuration section where you can switch between v3 and v4. Default is "autodetect" which tries to determine which api-endpoint is available (v3 is preferred for now). The result of the autodetection is cached after the first request to the server.

Additionally, i've tried to change the name of the classes so that the class-names correspond with the names of the getters and variables, thus GitLabApi is now called GitLabClient.

This is just some fancy code to switch the interface-definitions used to create a gitlab-rest-client. i'm not sure if switching between apis will work for existing jobs which do not have both the id and iid of merge-requests stored somewhere - could be that we have to change other parts to build a seamless upgrade path to v4.

ZJvandeWeg committed:
Upgrade to use GitLab APIv4

This commit is the simplest change which seems to suffice to move the
Jenkins plugin to APIv4. [The docs][api-docs] regarding the move from
v3 to v4 and going through most of the source, suggest that is the only
change too.

To summerize, the path changed from `api/v3` to `api/v4` and when
referencing merge requests in paths the IID is now needed instead of the
ID.

Tests are passing, but in this case I feel like that doesn't suggest
anything and this branch will undergo some manual testing soon.

[api-docs]: https://docs.gitlab.com/ce/api/v3_to_v4.html

Commit:
added support for travis-builds and maven-wrapper

Commit:
add docker-images to travis-cache

Commit:
fixed typo

Commit:
another try to cache docker-images

Commit:
try to execute unit- and integration tests in parallel using build-matrix

Commit:
fix travis.yml

Commit:
fix travis.yml another way

Commit:
disable findbugs for integration-tests

Commit:
always disable javadoc-generation for verification

Commit:
moved steps to separate shell-scripts

Commit:
cosmetics

Commit:
adds support for multiple gitlab-api-implementations and autodetects which api to use

Commit:
Merge branch 'gitlab-api-v4' into zj-upgrade-gitlab-api-v4

# Conflicts:
#	src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/V3GitLabClientProxy.java

Commit:
updated tests so both v3- and v4-apis are tested

bikebilly commented:
@Argelbargel that is a very good approach! Can't wait to see it released.

Argelbargel commented:
@bikebilly: if you're willing to install the plugin manually, you can try out a development-release from here: https://github.com/Argelbargel/gitlab-plugin/releases/gitlab-v4-SNAPSHOT/

Commit:
Merge remote-tracking branch 'upstream/master' into gitlab-api-v4

# Conflicts:
#	src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/V3GitLabClientProxy.java
#	src/test/java/com/dabsquared/gitlabjenkins/testing/gitlab/rule/GitLabRule.java
#	src/test/java/com/dabsquared/gitlabjenkins/testing/integration/GitLabIT.java

omehegan commented:
@Argelbargel this is great, thanks for taking the initiative to extend the original feature PR!

For the record, I understand that GitLab has decided they are not quite ready to deprecate the v3 API, so it will still work at least in the 10.0 release.

Argelbargel commented:
That's my understanding too - so we have enough time for a thorough review and perhaps might find some users to try out dev-snapshots in real-world environments.

bikebilly commented:
@omehegan API v3 have been deprecated back in GitLab 9.0, but until actual removal they will continue to work as usual.

Removal was discussed to happen in 10.0, but we decided to postpone so users have more time to update to the new version of the plugin (and other reasons unrelated to Jenkins).

Anyone interested can follow the plan at https://gitlab.com/gitlab-org/gitlab-ce/issues/36819.

Commit:
Merge branch 'upstream-master' into gitlab-api-v4

# Conflicts:
#	src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/V3GitLabClientProxy.java
#	src/test/java/com/dabsquared/gitlabjenkins/testing/gitlab/rule/GitLabRule.java
#	src/test/java/com/dabsquared/gitlabjenkins/testing/integration/GitLabIT.java

Commit:
updated new code from master

Commit:
Merge remote-tracking branch 'origin/gitlab-api-v4' into gitlab-api-v4

Argelbargel commented:
Has this any chance of being accepted anytime soon? At the current pace of changes in master, the effort of merging and fixing conflicts kinda nears the point where it'll be easier to start over when support for v4 is really required. No offense meant, it's partly my own fault, but nonetheless...

bikebilly commented:
@Argelbargel @omehegan if you don't have any strong reasons for not merging it, please do as soon as possible. We decided to postpone removal of API v3 mainly because of this plugin, and if you wait to merge since API are still available in the next version, we are in a deadlock. We have strong pressure on removing API v3, so I kindly ask you if we can move forward here. Please let us know if we can be of any help in the process. Thanks.

mreichel commented:
The pr looks pretty good to me. @Argelbargel could you look into the merge conflicts?

Argelbargel commented:
@mreichel: i did. that was the cause for my comment - as both mine and the current merges in master heavily change/rename the same classes i got a litte frustrated ;-)

I'll look further into it, it seems many of the conflicts occur because of changes in intendation/import ordering and such funny little things... is there a ruleset for formating for this project which contributors can use?

omehegan commented:
@Argelbargel I'm actively working on a release for the plugin this week, hence the merge conflicts. I apologize for that, I can see now that a better practice will be to review and merge PRs as they come in, rather than doing them in batches before a release. I'll do that going forward.

You've made some great contributions to the plugin so far, if you're interested I can add you as a committer and you can help out with reviews and merges. I would love to have more support! (I hope this doesn't sound snarky in response to your comment, it isn't meant that way!)

@bikebilly as soon as we can get this PR ready to merge cleanly, I will do that and release it. I apologize for the long delay lately, my free time to work on the plugin is sometimes very limited by other life obligations.

Argelbargel commented:
@omehegan: i think i'm already added as commiter/maintainer because of the (still pending :-/) work on the branch-source-branch. So if you'd like some help, i'll see what i can do. I propose to do some more reviews first and still let you do the merging so as to ensure what should be merged and what not..

It's totally understandable, that prs pile on, i just had to merge and fix conflicts a for this one a couple of times now and the last merges seem to take somewhat longer to integrate - thus my frustrated comment. 

I'll see that i'll fix the conflicts sunday or early next week so that we can review this and merge it before the next batch of changes comes in ;-)

Have a nice weekend!

omehegan commented:
@Argelbargel OK, I am going to release master today, the other open PRs are not ready to merge yet IMO. I will hold on merging anything else until this PR is ready to go, and then we can release this change by itself. I think that's wise in case we find issues with it, folks will be able to simply downgrade one version to remove this.

christ66 commented on src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnection.java:
```diff
@@ -25,15 +40,30 @@
 
     @DataBoundConstructor
-    public GitLabConnection(String name, String url, String apiTokenId, boolean ignoreCertificateErrors, Integer connectionTimeout, Integer readTimeout) {
+    public GitLabConnection(String name, String url, String apiTokenId, String clientBuilderId, boolean ignoreCertificateErrors, Integer connectionTimeout, Integer readTimeout) {
```
👎 as this breaks backward compatibility. The old constructor should still exist, and call the new constructor.

Argelbargel replied:
@christ66 you're right, this will break backward-compatiblity. But only in those case where someone outside of this plugin creates a GitLabConnection programmatically.

Is this class  intentionally part of the public api of this plugin (and not just because jenkins needs it to be to do it's magic)?  I'd say the only ways for anyone "outside" of this plugin to get a connection/api-client should be via GitLabConnectionProperty#getClient() or by looking up the configured connections by enumerating the GitLabConnectionConfigs and then calling GitLabConnectionConfig#getClient(). 

So instead of making an effort to keep compatibility here, i'd rather annotate at least the constructors with Restricted(NoExternalUse.class) so that they're only accessible outside of the plugin.

Or is there something i'm missing?

Oh, i've already added code to GitLabConnection#readResolve that ensures existing connections are upated to use the autodetecting client.

christ66 replied:
> You're right, this will break backward-compatiblity. But only in those case where someone outside of this plugin creates a GitLabConnection programmatically.

Right and that's the problem. There's no guarantee that someone has not created a groovy script which creates a Jenkins jobs and attaches a `GitLabConnection` using the old constructor. If you were to remove this constructor then that would mean that the person must update their scripts which might take some time, or they would have to do more work to figure out what happened. 

Here is an example where someone uses the constructor you're going to remove: 

https://github.com/rabits/jenkinsbro/blob/d5d279e340de638d6e5586292fd04ec5456e66a0/jenkinsBro/lib/modules/gitlab.groovy#L9-L14

This is just one example but someone might write a script for their non-public groovy scripts which will break, and that would mean that they need to update their scripts.

I would like to merge the PR which uses `@DataBoundSetter` annotations as this creates one empty constructor and everything is added using setter methods.

> Is this class intentionally part of the public api of this plugin (and not just because jenkins needs it to be to do it's magic)?

Once a method becomes public it's usually difficult to change this. It has nothing to do with any magic, but it's how public apis are. Once they're public it should be deprecated if they shouldn't be used.

Argelbargel replied:
So i'll keep the old constructor, deprecate it and move the @DateBoundConstructor-annotion to the new constructor (as there may only be one) and add a restricting annotion the prevent it leaking into the public api - except these constructors should be part of the public api.

Btw: with "jenkins-magic" i mean, that the class must be public because otherwise jenkins is not able to detect/call the @DataBoundConstructor.

bikebilly commented:
@omehegan please consider that we really appreciate your work, and we fully understand the situation. I just want to be sure that we have a shared plan, and that all our efforts are aimed to the same goal :smile:

Commit:
Merge branch 'upstream-master' into gitlab-api-v4

# Conflicts:
#	src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/ResteasyGitLabClientBuilder.java
#	src/main/java/com/dabsquared/gitlabjenkins/gitlab/api/impl/V3GitLabClientProxy.java
#	src/main/java/com/dabsquared/gitlabjenkins/util/CommitStatusUpdater.java

Commit:
- merged and massaged changes to master to work with our changes
- re-added old constructor to keep compatiblity

Commit:
- removed annotations for deprecation and restrictions (this pr might not be the right one to restrict currently public apis)

Argelbargel commented:
@christ66, @omehegan could you have another look? i'll have time tonight to make some more changes if required...

omehegan commented:
I'm leaving this for @christ66 to review, it's too far over my head and I trust him :)

christ66 commented:
Will review this weekend.

omehegan commented:
@christ66 any chance you will be able to look at this soon?

Argelbargel commented:
Hi guys, what about this: i'll deploy a snaphot-release of this tonight so that people get a chance to test it out. If there are no major complaints over the weekend, we at least merge this.

Then we can wait some more for further tests/reviews before releasing this but i cannot use it to procrastinate my work on #499 any longer ;-)

Argelbargel commented:
@omehegan: actually, i do not seem to know how to deploy a snapshot :-/ Which credentials do i have to pass to `mvn deploy` and how?

omehegan commented:
@Argelbargel I was thinking this exact thing this morning. I don't think anyone else but me is actually set up to push snapshots. Let me do that now and post the link.

omehegan commented:
http://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/gitlab-plugin/1.4.9-SNAPSHOT/gitlab-plugin-1.4.9-20170921.213040-1.hpi is built from this branch, ready for anyone to test.

Alexxtheonly commented:
The builds don't show up in the CI/CD menu of gitlab. Logs end with this entry:
```
INFORMATION: Project #2878 main build action completed: SUCCESS
Sep 26, 2017 3:40:57 PM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
```

Argelbargel commented:
@Alexxtheonly : thanks for your feedback. I'll check this myself, too but might not have time until thursday so perhaps you could try this to further localize the problem: do you use the v3 oder v4 api? If you did not change your settings then please disable autodetection. Go to the advanced section of the gitlab-connection settings and explicitly switch the api v3 or v4. If v3 still works than we've got another change to the api, otherwise i've broken something...

Alexxtheonly commented:
@Argelbargel I did change my settings because the old weren't working anymore.  
We used https as gitlab url. But that did not work 
```
Sep 27, 2017 6:54:55 AM org.apache.http.impl.client.DefaultRequestDirector tryConnect
INFORMATION: I/O exception (java.net.SocketException) caught when connecting to {s}->https://company.com:443: Connection reset
```

Using only `http` works fine.

**'Enable authentication for '/project' end-point' activated**

**v3**:
```
Sep 27, 2017 6:58:59 AM com.dabsquared.gitlabjenkins.trigger.handler.AbstractWebHookTriggerHandler handle
INFORMATION: Project triggered for push.
Sep 27, 2017 7:00:38 AM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
Sep 27, 2017 7:13:52 AM hudson.model.Run execute
INFORMATION: Project #2881 main build action completed: SUCCESS
Sep 27, 2017 7:14:00 AM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
Sep 27, 2017 7:15:07 AM hudson.model.Run execute
```

**v4**:
```
Sep 27, 2017 7:21:46 AM com.dabsquared.gitlabjenkins.trigger.handler.AbstractWebHookTriggerHandler handle
INFORMATION: Project triggered for push.
Sep 27, 2017 7:22:01 AM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
Sep 27, 2017 7:30:54 AM hudson.model.Run execute
INFORMATION: Project #2882 main build action completed: SUCCESS
Sep 27, 2017 7:30:58 AM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
Sep 27, 2017 7:31:51 AM hudson.model.Run execute
```
**'Enable authentication for '/project' end-point' deactivated**

**v3**:
```
Sep 27, 2017 7:46:01 AM com.dabsquared.gitlabjenkins.trigger.handler.AbstractWebHookTriggerHandler handle
INFORMATION: Project triggered for push.
Sep 27, 2017 7:46:09 AM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
```
**v4**:
```
Sep 27, 2017 7:34:48 AM com.dabsquared.gitlabjenkins.trigger.handler.AbstractWebHookTriggerHandler handle
INFORMATION: Project triggered for push.
Sep 27, 2017 7:35:03 AM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
Sep 27, 2017 7:43:58 AM hudson.model.Run execute
INFORMATION: Project #2883 main build action completed: SUCCESS
Sep 27, 2017 7:44:04 AM com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
INFORMATION: Retrieving gitlab project ids
```

Outcome is the same for all variants. No status reported back.
Any other things i could test without much effort?

Argelbargel commented:
@Alexxtheonly: none that i can think of :-/ I'll try to localize the problem later this week and cover it with tests. Though i do not have the faintest idea why this should also break https support  :-(

omehegan commented:
@Alexxtheonly FWIW, the "enable authentication for /project endpoint" feature should be irrelevant to the v3/v4 API changes. That just controls whether GitLab has to auth with Jenkins before it can trigger a build. It's unrelated to whether Jenkins can successfully send information to GitLab via their API.

Alexxtheonly commented:
@Argelbargel Any updates on this?

omehegan commented:
@Argelbargel @Alexxtheonly have you guys made any progress on this? Would it help for me to do some manual testing? I am anxious to get this feature released because it is holding up GitLab and also further development on the plugin at the moment. Maybe @christ66 or @mreichel could take a look as well?

Argelbargel commented:
No, not yet, sorry. I‘ll try to find some time to look into this tomorrow...

Argelbargel commented:
@Alexxtheonly,

i did finally find some time to look into this.

To check whether the gitlab-api has changed or there's something wrong with the new implementation  i've created a really simple freestyle-project which just listens to push events, checks out the commit and then publishes the build status to gitlab via the post-build action. At least my newly installed test-instance of gitlab 10.0.3 correctly shows the build-state (as pending while running and as passed afterwards) in the pipeline-view :-/

Does the freestyle-project match your use-case or are you using a Jenkinsfile and publish the build-status via the pipeline-step?

Which gitlab-version are you using? Then i'll try to check against your version too.

Alexxtheonly commented:
@Argelbargel thank you!

We are currently using 10.0.1
Both projects, one maven the other free style, built by jenkins did not report their status to gitlab. 

I'm currently on vacation, can test with the latest gitlab version next week.

bikebilly commented:
Hi, there should be no changes in API on GitLab side, in case let me know if we can help.

omehegan commented:
@Argelbargel have you tried using the docker config that is including in the plugin repo to set up a local test env? I moved it to the docker/ subdir and updated the README so it should give an OK guide on how to get started with it.

Argelbargel commented:
@omehegan: no, I'm using a vagrant-based setup we use to develop our pipelines and stuff.

Should not make much difference, though, As the docker images just contain gitlab with some test-projects, or am I mistaken?

Nonetheless, it might be a good idea to write/enhance an integration-test to run against the dockerized gitlab to flush out the cause for the problem. Shouldn't be too hard to do. I'll try that tomorrow.

omehegan commented:
@Argelbargel the docker setup does not even include test jobs, you would need to create some. I just thought that might help you try to replicate the issue @Alexxtheonly is seeing.

Alexxtheonly commented:
@Argelbargel I've updated gitlab to 10.0.3, unfortunately the issue persists.

Argelbargel committed:
Merge branch 'travis-support' into gitlab-api-v4

Argelbargel committed:
enforce ubuntu precise as distribution

Argelbargel committed:
added integration-test to ensure pipeline-state gets correctly reported to gitlab

Argelbargel committed:
removed travis-scripts again

Argelbargel committed:
Added the scripts again as the jenkins-based-pipeline seems not to run the integration-tests

Revert "removed travis-scripts again"

Argelbargel committed:
Added the scripts again as the jenkins-based-pipeline seems not to run the integration-tests

Revert "removed travis-scripts again"

Argelbargel committed:
- increased default-gitlab version to latest 8.x release (which supports pipelines-api required by new integration test)
- synchronized versions in docker-compose.yml with those in pom.xml to prevent errors

Argelbargel committed:
- further synchronized docker-compose.yml with settings in pom.xml to prevent errors

Argelbargel committed:
- increased postgres-version again

Argelbargel committed:
Update pom.xml

Argelbargel committed:
upgraded travis-build to stages (which should execute integration-tests for different gitlab versions)

Argelbargel committed:
travis-build corrected

Argelbargel committed:
travis-build corrected

Argelbargel committed:
skip installation-step

Argelbargel committed:
simplified build

Argelbargel committed:
cosmetics

Argelbargel committed:
let's see whether it's possible to use docker just for the integration-test stages...

Argelbargel committed:
hrmpf. okay. removed global sudo/docker

Argelbargel commented:
@Alexxtheonly i've written an integration-test for the problem - and it seems to me that build-status-reporting basically works. At least the pipeline shows up with the correct status and is queryable by the api. I can confirm, that the external jobs jenkins creates by reporting the build-status do not seem to show up on the jobs page - when you click on the pipeline-status the show up at the bottom under the heading "external", though.

@omehegan: as the jenkins-based checks seem not to run the integration-tests, i've re-added my travis-build-files for now. Would it be possible to run the integration-tests via jenkins too?

The results of the integration tests can be seen here: https://travis-ci.org/Argelbargel/gitlab-plugin 

For now i am at a loss what going wrong for you @Alexxtheonly that the pipeline in gitlab does not get updated. Might well be that i'm not testing the correct functionality - could you perhabs post some screenshots before (with the current gitlab-plugin which updates correctly) and after (with the snapshot-version of the plugin)? Perhaps this makes it easier for me to find out what exactly is not working anymore...

Alexxtheonly commented:
@Argelbargel Here are some screenshots:  

  
Since the gitlab update to 10.x the builds don't show up anymore, thought the `gitlab-plugin/1.4.9-SNAPSHOT/gitlab-plugin-1.4.9-20170921.213040-1.hpi` might fix that. That's how I found this pull request.

(The yellow minion is our jenkins user)
![image](https://user-images.githubusercontent.com/5155072/31725379-86a133ce-b424-11e7-87f8-dc155c653b4c.png)

Jenkins builds did never show up in 'jobs', so i guess it's ok.

Unfortunately I can not post a screenshot with working gitlab-plugin, because downgrading gitlab broke something i can't remember anymore(sorry).

That's all i can provide for now, at least till tomorrow.

Argelbargel commented:
Okay, just to get this right: the builds do not show up with the currently released version too? So while a problem it's not a regression caused by the changes in this pull-request?

Alexxtheonly commented:
Yes, since we updated gitlab to 10.x we do not see any builds. I thought the reason for this was the removed support for APIv3.

bikebilly commented:
@Alexxtheonly in GitLab 10.x we still support APIv3 (even if deprecated in favor of APIv4), so that is unlikely to be the problem: https://about.gitlab.com/2017/09/22/gitlab-10-0-released/#api-v3.

We removed CI API v1 (used by GitLab Runner < 9.0), but I don't think the plugin is relying on that.

Argelbargel commented:
@bikebilly: no, as far as i know, the plugin does not use the CI API.
@omehegan: if the problem reported by @Alexxtheonly is not a regression caused by the changes made in this PR, i propose that we merge this quite soon and build at least another snaphot the get more feedback and unblock the other PRs waiting on this.
@Alexxtheonly: i think you should open a separate issue for your problem as it is not related or at least not caused by the changes made in this PR. If you could perhaps add a simple project setup with which to reproduce your problem, that would be a great help. Perhaps add the versions of the plugins you're using, too. If your problem is not caused by your gitlab-instance (did you check the gitlab-logs for errors?), i think this could be caused by some underlying library like jackson-xml or so.

omehegan commented:
@Argelbargel can you squash your changes? I will release another snapshot shortly.

omehegan commented:
http://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/gitlab-plugin/1.4.9-SNAPSHOT/gitlab-plugin-1.4.9-20171018.215942-2.hpi has been released with the latest changes in this branch. Please test!

Alexxtheonly commented:
Did some further digging, our issues have nothing to do with this pull request. I'm sorry for any inconvenience caused. 


The issue for us is, when configuring 'Gitlab host URL' it fails when `https://` is used. But works with `http://`.
Log:
`I/O exception (java.net.SocketException) caught when connecting to {s}->https://gitlab.accenon.de:443: Connection reset`

Currently I don't know what exactly is causing this, but quick searches suggest it is related to our certificate (not self signed).

With http:// the connection check is successful, however, gitlab reports 301 as status. 
```
Got response from GitLab:
URL: http://domain.com/api/v3/projects/someProject/statuses/fe779b3782d0d7d3891c4e6b4d2e110756e0c7a7
Status: 301 Moved Permanently
```

Again sorry for any inconvenience caused. I will come back to report my test results regarding this pull request when I've resolved the other issue.

Argelbargel commented:
@omehegan: will do - or couldn't you just use the "squash and merge option" when merging this into master (click on the down-arrow right to "Merge pull request")?

Should i leave the travis-scripts in here or will you update the jenkins build to execute the integration-tests?

Alexxtheonly commented:
With help from @Argelbargel and a lot of trial and error, jenkins and gitlab are obeying my commands again :)

My tests in regards to this PR were all successful. I can confirm it works without any issues. 
v3, v4 and autodetect.

omehegan commented:
@Argelbargel oh I've never tried to squash-merge in GitHub, but OK! I can get Jenkins to run the additional tests (I think), let me give that a try.

@christ66 @mreichel would either of you like to give this PR a last check before we merge it?

Argelbargel commented:
@omehegan: that would be great. Don't think running the tests against Gitlab 8.x is that important but integrating against 9.x and upwards would be good to catch the differences to come...

mreichel reviewed (commented):
The last changes mainly affect the build process. I can't run the integration test profile on OSX or Linux because the Docker plugin is giving me some trouble. But I don't think that is a show stopper.

Argelbargel commented:
Hi guys, any feedback on this? There is a really long weekend coming up here in germany, so I'll have time to put some work into this if required.

omehegan commented:
@Argelbargel it sounds like @mreichel is happy, I didn't realize that the only real changes since her review and @christ66's were with the Travis build stuff. So I think we should go ahead and merge, and release this as a one-change release. I will do that this week, and we should ask GitLab to encourage their users to upgrade and test after we do that, just because they have the most frequent direct contact with users of this plugin.

omehegan commented:
OK, plugin version 1.5.0 has been released with only this change. It should be available in the update center within a few hours. Thanks everyone for your help getting this feature completed! If you discover any potential problems, please open new issues for them.

bikebilly commented:
Thanks everyone for your great job! We'll give the good news to our customers so they can upgrade to the new version.

Is APIv4 the default, or do you need to force them in the configuration? I'm wondering if upgrading the plugin is enough to test APIv4 calls, as we still support both versions.

Argelbargel commented:
@bikebilly : the default setting is "autodetect" which uses v3 when available and switches to v4 otherwise. The result of the autodetection is cached so it might be neccessary to restart jenkins after an upgrade to gitlab 11. Or we could easily change the order in which the apis are detected.
If users want to enforce the usage of v4 they can do so by setting the api-level via the advanced gitlab-configuration-options.

I'll provide an implementaiton of the autodetection which will not require a jenkins restart as soon as i can get my hands on a docker-image for gitlab 11 from https://github.com/sameersbn/docker-gitlab/releases - for now i chose this, admittedly a bit cumbersome variant, as it has the least possibility to break older setups.

bikebilly commented:
Oh I didn't see it coming early, sorry. That could be a problem as all the users that upgrade to the new plugin version will not test the APIv4 implementation (as we still support APIv3). This will not allow "involuntary" testing, but maybe it is a safer option.

What about this plan:
- we'll ask some customers to manually set APIv4 usage, and have some tests for a while
- if no problems are detected, you can change the plugin to make APIv4 the default option when autodetect
- APIv4 will be used by all users as they upgrade to this new version
- if no problems are spotted, we can release GitLab 11.0 without APIv3 support

It seems to me the best approach to avoid problems, and we actually don't already know when GitLab 11.0 will be out. Also, it is not strictly related to APIv3 removal, that can happen even later if it is really a necessity (but hopefully it will not be the case).

stefanandres commented:
I'm waiting to test this. I've upgraded the plugin to 1.5.0 already, how can I force it to use v4 in a multibranch pipeline job using the Jenkinsfile?

Argelbargel commented:
@bikebilly: good plan. and i'll try to re-implement the autodetection without requiring a restart right away ;-)

@stefanandres: you have to select it in the advanced-section for the gitlab-connection configuration in the system-wide-configuration
![gitlab-api-level](https://user-images.githubusercontent.com/54041/32050125-ffe2e256-ba4f-11e7-8b7b-234d33a8b906.png)
As long as you're not creating the gitlab-connection in your Jenkinsfile, the global settings should have effect for all jobs in the jenkins-instance

stefanandres commented:
Ah thanks, in the general settings. 

I can confirm that the v4 works as good as v3.

Though personally I still have the bug the runtime-status changes are not sent to gitlab. (that is something that worked with 1.3.0 or so and I've changed nothing in the configuration so far, as far as I know)

(Jenkinsfile: https://gist.github.com/stefanandres/7468a46b02267736a0cbb39c0622d356)
(log https://gist.github.com/stefanandres/c5f300463986a87232ec225c50cdb5f6)

mattiasb commented:
@stefanandres You need to wrap your code inside a `gitlabBuilds` step. Something like:
```groovy
gitlabBuilds(builds: ['build']) {
    gitlabCommitStatus('build') {
      stage name: 'Pre-Cleanup'
      deleteDir()
      […]
      }
    }
}
```
This is likely considered off-topic for this PR though so you should probably open a separate issue or ask somewhere else (no idea where).

bikebilly commented:
> @bikebilly: good plan. and i'll try to re-implement the autodetection without requiring a restart right away ;-)

@Argelbargel sounds very good to me. I already asked if we can select some customers for the first phase of the test for APIv4 support.

Argelbargel commented:
@bikebilly: i've changed the order in which the apis are detected and improved the autodetection. See #650 ;-)
